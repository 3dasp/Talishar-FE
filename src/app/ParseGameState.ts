import { isFunctionDeclaration } from 'typescript';
import Card from '../features/Card';
import CombatChainLink from '../features/CombatChainLink';
import GameState from '../features/GameState';
import CardDisplay from '../game/elements/CardDisplay';

const weaponArray = [
  'Pistol',
  'Sword',
  'Bow',
  'Staff',
  'Claw',
  'Hammer',
  'Dagger',
  'Gun',
  'Scepter',
  'Orb',
  'Axe',
  'Flail',
  'Scythe',
  'Club'
];

function ParseCard(input: any) {
  const card: Card = { cardNumber: 'blank' };
  card.cardNumber = input.cardNumber ? input.cardNumber : 'blank';
  card.action = card.action ? Number(card.action) : undefined;
  card.overlay = input.overlay == 1 ? 'disabled' : 'none';
  card.borderColor = input.borderColor ? String(input.borderColor) : undefined;
  card.actionDataOverride = input.actionDataOverride
    ? String(input.actionDataOverride)
    : undefined;
  card.counters = input.counters ? Number(input.counters) : 0;
  card.lifeCounters = input.lifeCounters ? Number(input.lifeCounters) : 0;
  card.defCounters = input.defCounters ? Number(input.defCounters) : 0;
  card.atkCounters = input.atkCounters ? Number(input.atkCounters) : 0;
  card.controller = input.controller ? Number(input.controller) : 0;
  card.type = input.type ? String(input.type) : undefined;
  card.sType = input.sType ? String(input.type) : undefined;
  card.restriction = input.restriction ? String(input.restriction) : undefined;
  card.isBroken = input.isBroken ? Boolean(input.isBroken) : false;
  card.onChain = input.onChain ? Boolean(input.onChain) : false;
  card.isFrozen = input.isFrozen ? Boolean(input.isFrozen) : false;
  card.gem = input.gem ? (input.gem == 2 ? 'active' : 'inactive') : 'none';

  return card;
}

export default function ParseGameState(input: any) {
  const result: GameState = {
    gameInfo: { gameID: 0, playerID: 0, authKey: '' },
    playerOne: {},
    playerTwo: {}
  };
  result.activeCombatChain = {};

  // combat Chain
  if (input.combatChain.length > 0) {
    result.activeCombatChain.attackingCard = ParseCard(input.combatChain[0]);
    if (input.combatChain.length > 1) {
      result.activeCombatChain.reactionCards = [];
      for (let i = 1; i < input.combatChain.length; i++) {
        result.activeCombatChain.reactionCards?.push(
          ParseCard(input.combatChain[i])
        );
      }
    }
  }
  result.activeCombatChain.totalAttack = input.totalAttack;
  result.activeCombatChain.totalDefence = input.totalDefence;

  // layers
  result.activeLayers = {};
  if (input.layerContents.length > 0) {
    result.activeLayers.active = true;
    result.activeLayers.cardList = [];
    for (const cardObj of input.layerContents) {
      result.activeLayers.cardList.push(ParseCard(cardObj));
    }
  } else {
    result.activeLayers.active = false;
  }

  // opponent hand
  result.playerTwo.Hand = [];
  for (const cardObj of input.opponentHand) {
    result.playerTwo.Hand.push(ParseCard(cardObj));
  }

  result.playerTwo.Health = input.opponentHealth;

  result.playerTwo.GraveyardCount = input.opponentDiscardCount;
  result.playerTwo.Graveyard = [];
  result.playerTwo.Graveyard.push(ParseCard(input.opponentDiscardCard));

  result.playerTwo.PitchRemaining = input.opponentPitchCount;
  result.playerTwo.Pitch = [];
  result.playerTwo.Pitch.push(ParseCard(input.opponentPitchCard));

  result.playerTwo.DeckSize = input.opponentDeckCount;
  result.playerTwo.DeckBack = ParseCard(input.opponentDeckCard);

  result.playerTwo.BanishCount = input.opponentBanishCount;
  result.playerTwo.Banish = [];
  result.playerTwo.Banish.push(ParseCard(input.opponentBanishCard));

  return result;
}
