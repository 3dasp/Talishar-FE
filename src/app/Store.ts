import { configureStore, ThunkAction, Action } from '@reduxjs/toolkit';
import gameReducer from '../features/game/GameSlice';
import type { PreloadedState } from '@reduxjs/toolkit';
import { apiSlice } from '../features/api/apiSlice';

export const store = configureStore({
  reducer: {
    game: gameReducer,
    [apiSlice.reducerPath]: apiSlice.reducer
  },
  middleware: (getDefaultMiddleware) =>
    getDefaultMiddleware().concat(apiSlice.middleware)
});

export const setupStore = (preloadedState?: PreloadedState<RootState>) => {
  return configureStore({
    reducer: {
      game: gameReducer,
      [apiSlice.reducerPath]: apiSlice.reducer
    },
    middleware: (getDefaultMiddleware) =>
      getDefaultMiddleware().concat(apiSlice.middleware),
    preloadedState
  });
};

export type AppDispatch = typeof store.dispatch;
export type AppStore = typeof store;
export type RootState = ReturnType<typeof store.getState>;
export type AppThunk<ReturnType = void> = ThunkAction<
  ReturnType,
  RootState,
  unknown,
  Action<string>
>;
