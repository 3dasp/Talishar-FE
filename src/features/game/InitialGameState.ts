import GameState from '../GameState';

const InitialGameState: GameState = {
  gameInfo: {
    gameID: 73333,
    playerID: 3,
    authKey: '28df413b665604299807c461a7f3cae71c4176cb2b96afad04b84cf96d016258'
  },
  activePlayer: 1,
  oldCombatChain: [
    {
      attackingCard: {
        cardNumber: 'WTR069'
      },
      reactionCards: [
        {
          cardNumber: 'CRU073'
        },
        {
          cardNumber: 'ARC044'
        }
      ],
      totalAttack: 88,
      totalDefence: 44,
      didItHit: true
    },
    {
      attackingCard: {
        cardNumber: 'WTR069'
      },
      reactionCards: [
        {
          cardNumber: 'CRU073'
        },
        {
          cardNumber: 'ARC044'
        }
      ],
      totalAttack: 88,
      totalDefence: 44,
      didItHit: true
    },

    {
      attackingCard: {
        cardNumber: 'WTR069'
      },
      reactionCards: [
        {
          cardNumber: 'CRU073'
        },
        {
          cardNumber: 'ARC044'
        }
      ],
      totalAttack: 88,
      totalDefence: 44,
      didItHit: true
    }
  ],
  activeCombatChain: {
    attackingCard: {
      cardNumber: 'CRU048'
    },
    reactionCards: [
      {
        cardNumber: 'CRU073'
      },
      {
        cardNumber: 'ARC044'
      }
    ],
    totalAttack: 88,
    totalDefence: 44,
    didItHit: true
  },
  playerOne: {
    // human player
    Name: 'LaustinSpayce',
    IsVerified: true,
    HeadEq: { cardNumber: 'WTR079' },
    ChestEq: { cardNumber: 'WTR150', counters: 2 },
    GlovesEq: { cardNumber: 'UPR158' },
    FeetEq: { cardNumber: 'WTR154' },
    WeaponLEq: { cardNumber: 'CRU048' },
    Hero: { cardNumber: 'CRU046' },
    WeaponREq: { cardNumber: 'CRU049' },
    Health: 20,
    ActionPoints: 0,
    PitchRemaining: 0,
    Permanents: [
      {
        cardNumber: 'ARC010',
        counters: 1
      },

      {
        cardNumber: 'ARC010'
      },
      {
        cardNumber: 'MON104'
      },
      {
        cardNumber: 'UPR409',
        lifeCounters: 5,
        counters: 12,
        defCounters: 873
      }
    ],
    Graveyard: [
      {
        cardNumber: 'ARC003'
      },
      {
        cardNumber: 'WTR069'
      },

      {
        cardNumber: 'WTR101'
      },
      {
        cardNumber: 'WTR100'
      },
      {
        cardNumber: 'WTR104'
      },
      {
        cardNumber: 'CRU073'
      },

      {
        cardNumber: 'WTR101'
      },
      {
        cardNumber: 'WTR100'
      },
      {
        cardNumber: 'WTR104'
      },
      {
        cardNumber: 'CRU073'
      },

      {
        cardNumber: 'WTR101'
      },
      {
        cardNumber: 'WTR100'
      },
      {
        cardNumber: 'WTR104'
      },
      {
        cardNumber: 'CRU073'
      },

      {
        cardNumber: 'WTR101'
      },
      {
        cardNumber: 'WTR100'
      },
      {
        cardNumber: 'WTR104'
      },
      {
        cardNumber: 'CRU073'
      },

      {
        cardNumber: 'WTR101'
      },
      {
        cardNumber: 'WTR100'
      },
      {
        cardNumber: 'WTR104'
      },
      {
        cardNumber: 'CRU073'
      },

      {
        cardNumber: 'WTR101'
      },
      {
        cardNumber: 'WTR100'
      },
      {
        cardNumber: 'WTR104'
      },
      {
        cardNumber: 'CRU073'
      }
    ],
    Pitch: [
      {
        cardNumber: 'ARC003'
      },
      {
        cardNumber: 'WTR069'
      }
    ],
    Hand: [
      {
        cardNumber: 'WTR101'
      },
      {
        cardNumber: 'WTR100'
      },
      {
        cardNumber: 'WTR104'
      },
      {
        cardNumber: 'CRU073'
      }
    ],
    Arsenal: [
      {
        cardNumber: 'UPR161'
      }
    ],
    Effects: [
      {
        cardNumber: 'ARC044'
      }
    ]
  },
  playerTwo: {
    // AI or opposing player
    Name: 'BigDumDum',
    HeadEq: { cardNumber: 'CRU006' },
    ChestEq: { cardNumber: 'WTR005' },
    GlovesEq: { cardNumber: 'WTR153' },
    FeetEq: { cardNumber: 'WTR004' },
    WeaponLEq: { cardNumber: '' },
    Hero: { cardNumber: 'WTR002' },
    WeaponREq: { cardNumber: 'WTR003' },
    Health: 20,
    ActionPoints: 0,
    PitchRemaining: 0,
    Hand: [
      {
        cardNumber: 'CBBlack'
      },
      {
        cardNumber: 'CBBlack'
      },
      {
        cardNumber: 'CBBlack'
      },
      {
        cardNumber: 'CBBlack'
      }
    ],
    Arsenal: [
      {
        cardNumber: 'CBBlack'
      }
    ],
    Banish: [
      {
        cardNumber: 'WTR104'
      },
      {
        cardNumber: 'CRU073'
      }
    ]
  },
  popup: {
    popupOn: false,
    popupCard: undefined
  }
};

export default InitialGameState as GameState;
