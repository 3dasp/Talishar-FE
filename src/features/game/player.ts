import { Card } from '../cardSlice';

export interface Player {
  HeadEq?: Card;
  ChestEq?: Card;
  GlovesEq?: Card;
  FeetEq?: Card;
  WeaponLEq?: Card;
  Hero?: Card;
  WeaponREq?: Card;
  Health?: number;
  ActionPoints?: number;
  PitchRemaining?: number;
  Hand?: Card[];
  Arsenal?: Card[];
  Banish?: Card[];
  Graveyard?: Card[];
  Pitch?: Card[];
  DeckSize?: number;
  Name?: string;
  IsVerified?: boolean;
}
