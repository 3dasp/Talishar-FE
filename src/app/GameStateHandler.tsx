import React, { useRef } from 'react';
import { useEffect } from 'react';
import { RootState } from './Store';
import {
  nextTurn,
  setGameStart,
  setIsUpdateInProgressFalse
} from '../features/game/GameSlice';
import { useAppDispatch, useAppSelector } from './Hooks';
import { shallowEqual } from 'react-redux';

export const GameStateHandler = React.memo(() => {
  const abortRef = useRef<AbortController>();
  const QueryParam = new URLSearchParams(window.location.search);
  const params = useAppSelector(
    (state: RootState) => state.game.gameInfo,
    shallowEqual
  );
  const isUpdateInProgress = useAppSelector(
    (state: RootState) => state.game.isUpdateInProgress
  );
  const dispatch = useAppDispatch();

  // setup long poll
  useEffect(() => {
    if (params.gameID == 0 || isUpdateInProgress) {
      return;
    }
    dispatch(nextTurn({ game: params, signal: abortRef.current?.signal }));
  }, [params, isUpdateInProgress, dispatch]);

  // write the gameID etc to the params
  useEffect(() => {
    let gameID = QueryParam.get('gameName') ? QueryParam.get('gameName') : '0';
    if (typeof gameID != 'string') {
      gameID = '0';
    }
    let player = QueryParam.get('playerID');
    if (player === null) {
      player = '3';
    }
    let authKey = QueryParam.get('authKey');
    if (!authKey) {
      // TODO: Get authKey from lastAuthKey cookie.
      authKey = '';
    }
    abortRef.current = new AbortController();

    dispatch(
      setGameStart({
        gameID: parseInt(gameID),
        playerID: parseInt(player),
        authKey: authKey
      })
    );

    return () => {
      abortRef.current?.abort();
      dispatch(setIsUpdateInProgressFalse());
    };
  }, []);

  return null;
});

GameStateHandler.displayName = 'GameStateHandler';
export default GameStateHandler;
