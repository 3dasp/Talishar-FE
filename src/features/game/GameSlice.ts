import { createAsyncThunk, createSlice, PayloadAction } from '@reduxjs/toolkit';
import ParseGameState from '../../app/ParseGameState';
import InitialGameState, { OfflineTestingGameState } from './InitialGameState';
import GameInfo from '../GameInfo';
import GameState from '../GameState';
import Card from '../Card';

export const nextTurn = createAsyncThunk(
  'game/nextTurn',
  async (params: GameInfo) => {
    // const queryURL = `http://localhost:41062/FaBOnline/GetNextTurn3.php?gameName=${params.gameID}&playerID=${params.playerID}`;
    const queryURL = `https://www.talishar.net/game/GetNextTurn3.php?gameName=${params.gameID}&playerID=${params.playerID}`;
    try {
      const response = await fetch(queryURL, {
        method: 'GET',
        headers: {}
      });
      const data = await response.text();
      const gameState: GameState = ParseGameState(data);
      return gameState;
    } catch (e) {
      console.error(e);
    }
  }
);

export const gameSlice = createSlice({
  name: 'game',
  // change the following line if you want to test with filled-in dummy data
  // initialState: InitialGameState,
  initialState: OfflineTestingGameState,
  // The `reducers` field lets us define reducers and generate associated actions
  reducers: {
    setPopUp: (
      state,
      action: PayloadAction<{
        cardNumber: string;
        xCoord?: number;
        yCoord?: number;
      }>
    ) => {
      state.popup = {
        popupOn: true,
        xCoord: action.payload.xCoord,
        yCoord: action.payload.yCoord,
        popupCard: { cardNumber: action.payload.cardNumber }
      };
    },
    clearPopUp: (state) => {
      state.popup = { popupOn: false, popupCard: undefined };
    },
    setPlayCardMessage: (state) => {
      state.playCardMessage = {
        popUpOn: true,
        message: 'Release to play this card'
      };
    },
    clearPlayCardMessage: (state) => {
      state.playCardMessage = { popUpOn: false };
    },
    setCardListFocus: (
      state,
      action: PayloadAction<{
        cardList?: Card[];
        name?: string;
      }>
    ) => {
      state.cardListFocus = {
        active: true,
        cardList: action.payload.cardList,
        name: action.payload.name
      };
    },
    clearCardListFocus: (state) => {
      state.cardListFocus = undefined;
    },
    setGameStart: (
      state,
      action: PayloadAction<{ playerID: number; gameID: number }>
    ) => {
      (state.gameInfo.gameID = action.payload.gameID),
        (state.gameInfo.playerID = action.payload.playerID);
    }
  },
  // The `extraReducers` field lets the slice handle actions defined elsewhere,
  // including actions generated by createAsyncThunk or in other slices.
  extraReducers: (builder) => {
    builder.addCase(nextTurn.fulfilled, (state, action) => {
      if (action.payload === undefined) {
        return state;
      }
      state.playerOne = action.payload.playerOne;
      state.playerTwo = action.payload.playerTwo;
      state.activeCombatChain = action.payload.activeCombatChain;
      state.activeLayers = action.payload.activeLayers;
      state.oldCombatChain = action.payload.oldCombatChain;
      state.chatLog = action.payload.chatLog;
      return state;
    });
  }
});

// export const {} = gameSlice.actions;

// We can also write thunks by hand, which may contain both sync and async logic.
// Here's an example of conditionally dispatching actions based on current state.

export default gameSlice.reducer;

const { actions } = gameSlice;
export const {
  setPopUp,
  setGameStart,
  clearPopUp,
  setCardListFocus,
  clearCardListFocus
} = actions;
