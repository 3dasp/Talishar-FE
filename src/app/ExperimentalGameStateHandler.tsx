import React, { useEffect } from 'react';
import { useCookies } from 'react-cookie';
import { useParams } from 'react-router-dom';
import { useAppDispatch, useAppSelector } from './Hooks';
import { getGameInfo, nextTurn } from 'features/game/GameSlice';
import { shallowEqual } from 'react-redux';

const ExperimentalGameStateHandler = () => {
  const [cookies] = useCookies(['experimental']);
  const { gameID } = useParams();
  const gameInfo = useAppSelector(getGameInfo, shallowEqual);
  const dispatch = useAppDispatch();

  useEffect(() => {
    console.log('setting up listener');
    const source = new EventSource(
      `https://api.talishar.net/game/GetUpdateSSE.php?gameName=${gameID}&playerID=${gameInfo.playerID}&authKey=${gameInfo.authKey}`
    );
    console.log(source);
    source.onmessage = (e) => {
      console.log('update data:', e.data);
      dispatch(
        nextTurn({
          game: gameInfo,
          signal: undefined,
          lastUpdate: 0
        })
      );
    };

    return () => {
      console.log('closing eventstream');
      source.close();
    };
  }, []);

  return null;
};

export default ExperimentalGameStateHandler;
