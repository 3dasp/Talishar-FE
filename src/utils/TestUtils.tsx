/* eslint-disable @typescript-eslint/ban-types */
import React, { PropsWithChildren } from 'react';
import { render } from '@testing-library/react';
import type { RenderOptions } from '@testing-library/react';
import type { PreloadedState } from '@reduxjs/toolkit';
import { Provider } from 'react-redux';

import { AppStore, RootState, setupStore } from '../app/Store';

import InitialGameState from '../features/game/InitialGameState';
import { MemoryRouter } from 'react-router-dom';

// This type interface extends the default options for render from RTL, as well
// as allows the user to specify other things such as initialState, store.
interface ExtendedRenderOptions extends Omit<RenderOptions, 'queries'> {
  preloadedState?: PreloadedState<RootState>;
  store?: AppStore;
}

export const renderWithProviders = (
  ui: React.ReactElement,
  {
    // @ts-ignore - I know I know...
    preloadedState = {
      game: InitialGameState
    },
    // Automatically create a store instance if no store was passed in
    store = setupStore(preloadedState),
    ...renderOptions
  }: ExtendedRenderOptions = {}
) => {
  function Wrapper({ children }: PropsWithChildren<{}>): JSX.Element {
    return (
      <Provider store={store}>
        <MemoryRouter>{children}</MemoryRouter>
      </Provider>
    );
  }
  return { store, ...render(ui, { wrapper: Wrapper, ...renderOptions }) };
};
