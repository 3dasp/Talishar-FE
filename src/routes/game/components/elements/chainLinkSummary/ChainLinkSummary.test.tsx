import React from 'react';
import { store } from 'app/Store';
import ChainLinkSummaryContainer from './ChainLinkSummary';
import { renderWithProviders } from 'utils/TestUtils';
import { setGameStart, showChainLinkSummary } from 'features/game/GameSlice';

it('renders without crashing isPlayer true', () => {
  const div = renderWithProviders(<ChainLinkSummaryContainer />);
  const display = document.querySelector('div');
  expect(display).toMatchSnapshot();
});

it('shows a mock chain link summary', async () => {
  const testStore = store;
  store.dispatch(showChainLinkSummary({ chainLink: 0 }));
  store.dispatch(
    setGameStart({ playerID: 1, gameID: 123456789, authKey: '12345' })
  );
  const div = renderWithProviders(<ChainLinkSummaryContainer />, {
    store: testStore
  });
  await div.findByText('Zipper Hit');
  const display = document.querySelector('div');
  expect(display).toMatchSnapshot();
});
