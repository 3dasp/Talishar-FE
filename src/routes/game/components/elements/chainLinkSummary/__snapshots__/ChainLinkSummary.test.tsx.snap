// Vitest Snapshot v1

exports[`renders without crashing isPlayer true 1`] = `<div />`;

exports[`shows a mock chain link summary 1`] = `
<div>
  <div>
    <div
      class="_emptyOutside_1e6509"
    >
      <div
        class="_cardListBox_1e6509"
      >
        <div
          class="_cardListTitleContainer_1e6509"
        >
          <div
            class="_cardListTitle_1e6509"
          >
            <h3
              class="_title_1e6509"
            >
              Chain Link Summary
            </h3>
          </div>
          <div
            class="_cardListCloseIcon_1e6509"
          >
            <svg
              fill="currentColor"
              height="1em"
              stroke="currentColor"
              stroke-width="0"
              viewBox="0 0 352 512"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <title>
                close dialog
              </title>
              <path
                d="M242.72 256l100.07-100.07c12.28-12.28 12.28-32.19 0-44.48l-22.24-22.24c-12.28-12.28-32.19-12.28-44.48 0L176 189.28 75.93 89.21c-12.28-12.28-32.19-12.28-44.48 0L9.21 111.45c-12.28 12.28-12.28 32.19 0 44.48L109.28 256 9.21 356.07c-12.28 12.28-12.28 32.19 0 44.48l22.24 22.24c12.28 12.28 32.2 12.28 44.48 0L176 322.72l100.07 100.07c12.28 12.28 32.2 12.28 44.48 0l22.24-22.24c12.28-12.28 12.28-32.19 0-44.48L242.72 256z"
              />
            </svg>
          </div>
        </div>
        <div
          class="_cardListContents_1e6509"
        >
          <div>
            Total Damage: 
          </div>
          <table
            class="_table_1e6509"
          >
            <thead>
              <tr>
                <th
                  class="_leftColumn_1e6509"
                />
                <th
                  class="_midColumn_1e6509"
                >
                  Card
                </th>
                <th
                  class="_rightColumn_1e6509"
                >
                  Effect
                </th>
              </tr>
            </thead>
            <tbody>
              <tr>
                <td
                  class="_column_1e6509"
                >
                  <div
                    class="_effect_159bb8"
                  >
                    <img
                      class="_img_159bb8"
                      src="/crops/ARC030_cropped.png"
                    />
                  </div>
                </td>
                <td
                  class="_column_1e6509"
                >
                  <span
                    class="_cardText_1b6faf"
                  >
                    Zipper Hit
                  </span>
                </td>
                <td
                  class="_column_1e6509"
                >
                  +
                  4
                </td>
              </tr>
            </tbody>
          </table>
        </div>
      </div>
    </div>
  </div>
</div>
`;
